src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs

```diff
@@ -6826,6 +6832,13 @@ private BoundExpression ConvertToArrayIndex(BoundExpression index, SyntaxNode no
                 TryImplicitConversionToArrayIndex(index, SpecialType.System_Int64, node, diagnostics) ??
                 TryImplicitConversionToArrayIndex(index, SpecialType.System_UInt64, node, diagnostics);
 
+            if (result is null && allowIndexAndRange)
+            {
+                result = 
+                    TryImplicitConversionToArrayIndex(index, WellKnownType.System_Index, node, diagnostics) ??
+                    TryImplicitConversionToArrayIndex(index, WellKnownType.System_Range, node, diagnostics);
+            }
+
             if (result == null)
```

Comment:
Can you make this `result is null` for consistency with the rest of this change?